### ISSUE TITLE
How do you implement clustering of face landmarks?

### ISSUE THREAD

**Issue by Adenialzz** [hysts/anime-face-detector#10]
Thank you for sharing this wonderful project. I am curious about how do you implement clustering of face landmarks. Can you describe that in detail? Or can you sharing some related papers or projects? Thanks in advance.

**Comment by hysts** [comment]
Hi, @Adenialzz 

I may have forgotten some details, but the following is the rough idea for clustering:
- collect enough images
- predict face boxes and landmarks
- crop faces with the box and compute the mean landmarks
- compute similarity transform that best fits landmarks to the mean landmarks
- align landmarks and images with the similarity transform
- treat the aligned landmarks as vectors and perform k-means clustering

Does this make sense to you?

**Comment by Adenialzz** [comment]
Hi, @hysts 
Thank you for your reply. I'm not very familiar with keypoints, so I'm not very clear about some concepts (mean landmarks, similarity transform) in your reply. Can you briefly introduce their practices (or share some related material).
I made some attempts according to my own understanding these days: after predicting the coordinates of 28 keypoints of all faces in the image library, I normalize them, i.e. (x_i-x_min)/(x_max-x_min),  (y_i-y_min)/(y_max-y_min) , and then treat the normalized 28 points as a 56 dimensional feature vector, and implement K-means clustering on all faces. Do you think it makes sense?

**Comment by hysts** [comment]
@Adenialzz 
It seems preprocessing is different from what I did, but I think that may work too. I wonder how it went.

As for the mean landmarks and similarity transform, what I meant was simply averaging the landmarks and using `skimage.transform.SimilarityTransform` to estimate similarity transform between the landmarks and the mean landmarks.

Anyway, I think it's easier to look at codes, so I'll search notebooks I used and share the code later if they're still there. It may take some time, though.

**Comment by hysts** [comment]
@Adenialzz 

Here is a cleaned-up code for clustering. This is not the exact same code that I used for creating images in README, but this should help you understand how the clustering is done.
```python
import pathlib

import cv2
import matplotlib.pyplot as plt
import numpy as np
import skimage.transform
import sklearn.cluster
import tqdm.auto
from anime_face_detector import create_detector
from mpl_toolkits.axes_grid1 import ImageGrid

crop_size = 256
# This file can be downloaded from https://github.com/hysts/anime-face-detector/releases/tag/v0.0.1
mean_pts = np.load('mean_pts.npy')
mean_pts = mean_pts * crop_size * 0.5 + np.array(
    [crop_size * 0.5, crop_size * 0.65])

image_paths = sorted(pathlib.Path('images').glob('*'))
detector = create_detector('yolov3')

# Align images and landmarks
aligned_images = []
aligned_pts = []
for image_path in tqdm.auto.tqdm(image_paths):
    image = cv2.imread(image_path.as_posix())
    preds = detector(image)
    for pred in preds:
        pts = np.round(pred['keypoints'])[:, :2].astype(int)

        sim_transform = skimage.transform.SimilarityTransform()
        sim_transform.estimate(mean_pts, pts)
        new_pts = np.hstack([pts, np.ones(
            (len(pts), 1))]) @ np.linalg.inv(sim_transform.params.T)
        new_pts = new_pts[:, :2]

        crop = skimage.transform.warp(image,
                                      sim_transform,
                                      output_shape=(crop_size, crop_size),
                                      order=3,
                                      mode='edge',
                                      preserve_range=True).astype(np.uint8)

        # Remove failed images
        new_preds = detector(crop)
        if len(new_preds) != 1:
            continue
        pts2 = np.array(new_preds[0]['keypoints'])[:, :2]
        diff = np.sqrt(np.sum((new_pts - pts2)**2, axis=1)).mean()
        if diff > 5:
            continue

        aligned_images.append(crop)
        aligned_pts.append(new_pts)
aligned_pts = np.array(aligned_pts)

# Apply k-means clustering
n_clusters = 9
kmeans = sklearn.cluster.KMeans(n_clusters=n_clusters, random_state=0)
kmeans.fit(aligned_pts.reshape(aligned_pts.shape[0], -1))

cluster_indices = [
    np.arange(len(aligned_pts))[kmeans.labels_ == i] for i in range(n_clusters)
]
cluster_indices.sort(key=lambda x: -len(x))

ncols = int(np.ceil(np.sqrt(n_clusters)))
nrows = int(np.ceil(n_clusters / ncols))

mean_pt_images = []
for i in range(n_clusters):
    indices = cluster_indices[i]
    mean_shape = aligned_pts[indices].mean(axis=0)
    res = np.zeros((crop_size, crop_size, 3), dtype=np.uint8)
    for pt in mean_shape:
        pt = np.round(pt).astype(int)
        cv2.circle(res, tuple(pt), 2, (0, 255, 0), cv2.FILLED)
    mean_pt_images.append(res)
for _ in range(nrows * ncols - n_clusters):
    mean_pt_images.append(np.zeros_like(mean_pt_images[0]))

fig = plt.figure(figsize=(15, 15))
grid = ImageGrid(fig, 111, nrows_ncols=(nrows, ncols), axes_pad=0.1)
for ax, image in zip(grid, mean_pt_images):
    ax.imshow(image[:, :, ::-1])
    ax.axis('off')
plt.show()

mean_images = []
for i in range(n_clusters):
    indices = cluster_indices[i]
    res = np.zeros((crop_size, crop_size, 3), dtype=np.float32)
    for i in indices:
        image = aligned_images[i]
        res += image.astype(np.float32)
    res = np.clip(np.round(res / len(indices)), 0, 255).astype(np.uint8)
    mean_images.append(res)
for _ in range(nrows * ncols - n_clusters):
    mean_images.append(np.zeros_like(mean_images[0]))

fig = plt.figure(figsize=(15, 15))
grid = ImageGrid(fig, 111, nrows_ncols=(nrows, ncols), axes_pad=0.1)
for ax, image in zip(grid, mean_images):
    ax.imshow(image[:, :, ::-1])
    ax.axis('off')
plt.show()
```

**Comment by Adenialzz** [comment]
Thanks a lot. I'll try your code after the weekend. By the way, the clustering implementation I mentioned earlier has also achieved passable results. I will discuss it with you after trying your implementation carefully. Thanks again.

**Comment by Adenialzz** [comment]
@hysts 
Hello, I have finished the experiment of two preprocessing methods, and now the conclusion is that your method is more reasonable.

I'm not familiar with how to utilize the keypoints of face, but my basic idea is that clustering needs to distinguish the features of face, such as orientation of face, the opening and closing of mouth/eyes and so on. And the result should have nothing to do with the size of image, the position of face in the original image. Therefore, I made a simple normalization (as I mentioned earlier) to eliminate the influence of these factors on the coordinate values of keypoints, and mapped the values to 0-1 for convenient calculation. After reading and trying your code implementation, I also read some materials and learned that "alignment" is a very common and useful operation in face keypoints processing. It can also eliminate the influence of the factors mentioned above (size, position) and place the face in the middle of the image of the same size, which also makes your visualization looks very fancy.

Previously, I mentioned that my simple preprocessing method has also achieved passable results. I mean that in my clustering results,  images from the same cluster have obvious features, such as the orientation of most faces is the same. However, because there is no alignment operation, the result looks very bad when using the same visualization as yours. Maybe I should find a more suitable visualization method for my method. Anyways, I decide to adopt your preprocessing method, which is more professional and more intuitive.
Now, I still have a few questions to discuss with you:

1. I noticed that in your visualization results, in addition to the obvious feature of face orientation, more detailed features such as the opening and closing of people's mouths/eyes can also be separated in clustering. But in my experiment, this result was not obviously observed. I guess there are several possible reasons:
  -Your dataset has sufficient samples of such characteristics, and the number of various clusters is relatively balanced;
  -Your choice of K-means initial point is exquisite
What reason do you think (maybe not among the reasons I listed) makes you get such a good result? I will refer to your reply and adjust my own data or code to get a similar effect.
2. If I want to explicitly add some features, such as the area of eyes, the area of mouth, or the proportion of their area to the total face area, and the length-width ratio of eyes and eyes, so as to get the clustering effect I want. Do you think this is feasible? Can I directly calculate the above values and concatenate them to the back of the keypoints coordinate feature vector, or is there any other more reasonable feature aggregation method?
3. Do you have any unique experience in the selection of K value and initial point? Does this require a lot of observation of the data distribution in the data set?
Thanks again for your patient reply and code sharing. I feel very happy to talk with you. And I realize that you have a deep research in the field of computer vision, so I have raised above questions and hope you can give me some guidance when you have time :) .

**Comment by hysts** [comment]
Hi, @Adenialzz 

So sorry for late reply. I just noticed your comment on this issue. It seems I had missed the notification.
Thanks for sharing your results. 

> 1. What reason do you think (maybe not among the reasons I listed) makes you get such a good result?

I think it's probably the first possibility you mentioned.

> 2. Do you think this is feasible? Can I directly calculate the above values and concatenate them to the back of the keypoints coordinate feature vector, or is there any other more reasonable feature aggregation method?

I think it could work. But you may want to normalize landmark values and your computed features. If the scale of the values the landmarks and your features are too different, I think it wouldn't work as expected.

> 3. Do you have any unique experience in the selection of K value and initial point? Does this require a lot of observation of the data distribution in the data set?

No, I don't. I've experimented with datasets consisting of normal human faces etc., though. If the clustering result is not what you expected, you may want to look at the images belonging to each cluster.

**Comment by Adenialzz** [comment]
> Hi, @Adenialzz
> 
> So sorry for late reply. I just noticed your comment on this issue. It seems I had missed the notification. Thanks for sharing your results.
> 
> > 1. What reason do you think (maybe not among the reasons I listed) makes you get such a good result?
> 
> I think it's probably the first possibility you mentioned.
> 
> > 2. Do you think this is feasible? Can I directly calculate the above values and concatenate them to the back of the keypoints coordinate feature vector, or is there any other more reasonable feature aggregation method?
> 
> I think it could work. But you may want to normalize landmark values and your computed features. If the scale of the values the landmarks and your features are too different, I think it wouldn't work as expected.
> 
> > 3. Do you have any unique experience in the selection of K value and initial point? Does this require a lot of observation of the data distribution in the data set?
> 
> No, I don't. I've experimented with datasets consisting of normal human faces etc., though. If the clustering result is not what you expected, you may want to look at the images belonging to each cluster.

**Comment by Adenialzz** [comment]
> Hi, @Adenialzz
> 
> So sorry for late reply. I just noticed your comment on this issue. It seems I had missed the notification. Thanks for sharing your results.
> 
> > 1. What reason do you think (maybe not among the reasons I listed) makes you get such a good result?
> 
> I think it's probably the first possibility you mentioned.
> 
> > 2. Do you think this is feasible? Can I directly calculate the above values and concatenate them to the back of the keypoints coordinate feature vector, or is there any other more reasonable feature aggregation method?
> 
> I think it could work. But you may want to normalize landmark values and your computed features. If the scale of the values the landmarks and your features are too different, I think it wouldn't work as expected.
> 
> > 3. Do you have any unique experience in the selection of K value and initial point? Does this require a lot of observation of the data distribution in the data set?
> 
> No, I don't. I've experimented with datasets consisting of normal human faces etc., though. If the clustering result is not what you expected, you may want to look at the images belonging to each cluster.

Thanks for your reply, I will try again with reference to your suggestion.
